### PR TITLE
[NEW] Service account creation

### PR DESCRIPTION
**Pull request by bhardwajaditya**
**Service Account Creation UI**

service-account-creation

**Service Account Login** 

service-account-login

- Added a method to update the service account owner username when the user updates his username.
- Pull request passed all CLI tests.

- Updated service accounts dashboard to approve only first service account for each user.
- Added API and test

### CODE REVIEW

**Comment by bizzbyster** [comment]
Remove "channels" line from gif

**Comment by kb0304** [comment]
Will review after #4 is merged.

**Comment by kb0304** [comment]
Reopening to refresh the diffs

File: app/lib/server/functions/setUsername.js [hunk 1]
```diff
@@ -79,6 +79,7 @@ export const _setUsername = function(userId, u) {
 		Subscriptions.setUserUsernameByUserId(user._id, username);
 		Subscriptions.setNameForDirectRoomsWithOldName(previousUsername, username);
 		LivechatDepartmentAgents.replaceUsernameOfAgentByUserId(user._id, username);
+		Users.setOwnerUsernameByUserId(user._id, username);
```
Comment (kb0304) [inline app/lib/server/functions/setUsername.js:82]: Shouldn't this be under a flag which indicates this is a Service account's username being modified.
Also, to run this function a user is required to have this permission `edit-other-user-inf`??

File: app/lib/server/functions/saveUser.js [hunk 2]
```diff
@@ -170,6 +172,9 @@ export const saveUser = function(userId, userData) {
 		if (userData.email) {
 			createUser.email = userData.email;
 		}
+		if (userData.u) {
+			createUser.u = userData.u;
```
Comment (kb0304) [inline app/lib/server/functions/saveUser.js:176]: Should there be some sort of authorisation check before this update? Should we allow a service account to update its owner?

File: app/service-accounts/server/methods/usernameExists.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { Meteor } from 'meteor/meteor';
+import { check } from 'meteor/check';
+
+import { Users } from '../../../models';
+
+Meteor.methods({
+	usernameExists(name) {
```
Comment (kb0304) [inline app/service-accounts/server/methods/usernameExists.js]: You might want to add rate-limiter to this method

File: app/lib/server/functions/saveUser.js (hunk 2 above)
Comment (bhardwajaditya) [reply app/lib/server/functions/saveUser.js:176]: This is not update. This is when a user creates new account. In this file they used if else condition to check when it's update and when it's new user. This userData.u is inserted when service account  is created.

File: app/lib/server/functions/setUsername.js (hunk 1 above)
Comment (bhardwajaditya) [reply app/lib/server/functions/setUsername.js:82]: This is called when owner username is modified. The same happens for subscription and room user info. This function will update only those service accounts where the current user is the owner.

File: app/service-accounts/server/methods/usernameExists.js (hunk 3 above)
Comment (bhardwajaditya) [reply app/service-accounts/server/methods/usernameExists.js]: When a user is creating a new service account, this is to show which username is not available before submission. Should I put a 10 calls per user per hour limit?

Comment (kb0304) [reply app/service-accounts/server/methods/usernameExists.js]: There is an already existing implementation of this, which also includes check for blocklisted usernames. Can you update the code to use that?
See app/lib/server/functions/checkUsernameAvailability.js

Comment (bhardwajaditya) [reply app/service-accounts/server/methods/usernameExists.js]: I have updated this method with this function.

**Review by kb0304 — APPROVED** [review]

**Pull request merged**